scipy/sparse/csgraph/_flow.pyx

```diff
@@ -668,3 +682,753 @@ cdef ITYPE_t[:] _dinic(
+    # Insert the subtree rooted at q into the depth-first thread.
+    next_vertex_dft[last_p] = q
+    prev_vertex_dft[q] = last_p
+    prev_vertex_dft[next_last_p] = last_q
+    next_vertex_dft[last_q] = next_last_p
+    # Update the subtree sizes and last descendants of the (new) ancestors
+    # of q.
+    while p != -2:
+        subtree_size[p] += size_q
+        if last_descendent_dft[p] == last_p:
+            last_descendent_dft[p] = last_q
+        p = parent[p]
+
+cdef inline bint _trace_subtree(
+        ITYPE_t p,  # IN
+        ITYPE_t[:] last_descendent_dft,  # IN
+        ITYPE_t[:] next_vertex_dft,  # IN
+        ITYPE_t[:] subtree  # IN/OUT
+        ) nogil:
+    cdef ITYPE_t idx, l
+    subtree[0] = p
+    idx = 1
+    l = last_descendent_dft[p]
+    while p != l:
+        p = next_vertex_dft[p]
+        subtree[idx] = p
+        idx += 1
+    return idx
+
+cdef void _update_potentials(
+        ITYPE_t i,  # IN
+        ITYPE_t p,  # IN
+        ITYPE_t q,  # IN
+        ITYPE_t[:] edge_targets,  # IN
+        ITYPE_t[:] edge_weights,  # IN
+        ITYPE_t[:] last_descendent_dft,  # IN
+        ITYPE_t[:] next_vertex_dft,  # IN
+        ITYPE_t[:] vertex_potentials,  # IN/OUT
+        ITYPE_t[:] subtree,  # IN/OUT
+        ) nogil:
+    cdef ITYPE_t idx, d, path_len
+
+    if q == edge_targets[i]:
+        d = vertex_potentials[p] - edge_weights[i] - vertex_potentials[q]
+    else:
+        d = vertex_potentials[p] + edge_weights[i] - vertex_potentials[q]
+    path_len = _trace_subtree(q, last_descendent_dft,
+                              next_vertex_dft, subtree)
+    for idx in range(path_len):
+        q = subtree[idx]
+        vertex_potentials[q] += d
+
+cdef inline void _add_entry(
+        ITYPE_t edge_source,  # IN
+        ITYPE_t edge_target,  # IN
+        ITYPE_t flow,  # IN
+        ITYPE_t idx,  # IN
+        ITYPE_t[:] row,  # IN/OUT
+        ITYPE_t[:] col,  # IN/OUT
+        ITYPE_t[:] flow_data  # IN/OUT
+        ) nogil:
+    row[idx] = edge_source - 1
+    col[idx] = edge_target - 1
+    flow_data[idx] = flow
+
+cdef ITYPE_t[:] _network_simplex(
+        ITYPE_t[:] heads,  # IN
+        ITYPE_t[:] tails,  # IN
+        ITYPE_t[:] capacities,  # IN
+        ITYPE_t[:] demand,  # IN
+        ITYPE_t[:] cost,  # IN
+        ITYPE_t n_verts,  # IN
+        ITYPE_t[:] row,  # IN/OUT
+        ITYPE_t[:] col,  # IN/OUT
+        ITYPE_t[:] flow_data  # IN/OUT
+        ):
+    cdef ITYPE_t n_edges = capacities.shape[0]
+    cdef ITYPE_t idx, faux_inf, capacities_sum, cost_sum
+    cdef ITYPE_t j, s, t, i, tmp, p, q, n_non_zero_edges, v
+    cdef ITYPE_t flow_cost, flow_value
+    cdef ITYPE_t[:] edge_flow, vertex_potentials, Wne_len
+    cdef ITYPE_t[:] parent, parent_edge, subtree_size
+    cdef ITYPE_t[:] next_vertex_dft, prev_vertex_dft
+    cdef ITYPE_t[:] last_descendent_dft
+    cdef ITYPE_t[:] Wn, We, WnR, WeR, ancestors, subtree
+    cdef local_vars *B_M_m_f
+    cdef edge_result *i_p_q
```

Comment:
It should be possible to have to have those wrappers be stack-allocated, thus getting ride of `malloc` and using:
```suggestion
    cdef local_vars B_M_m_f
    cdef edge_result i_p_q
```